ADT7310 SPI temperature sensor driver

gebart opened the pull request:
drivers/adt7310: Add ADT7310 temperature sensor driver
    
This is a driver for Analog Devices ADT7310 +/-0.5°C Accurate, 16-Bit Digital SPI Temperature Sensor.

Only basic reading is supported. Interrupt and compare are not implemented yet.
    
See: http://www.analog.com/en/products/analog-to-digital-converters/integrated-special-purpose-converters/integrated-temperature-sensors/adt7310.html

Tested on mulle

LudwigOrtmann commented:
In case you don't want to implement interrupt mode etc. as part of this PR you could document this missing feature in the API.. Compare http://riot-os.org/api/group__driver__isl29125.html#details

gebart commented:
@LudwigOrtmann Thank you for your suggestion. I have updated the PR with a little more documentation for the main page.

gebart commented:
I fixed the cppcheck warning and removed the extra whitespaces that showed up in Travis. Waiting for a new Travis run.

LudwigOrtmann commented on drivers/include/adt7310.h:
```diff
@@ -0,0 +1,164 @@
+int adt7310_init(adt7310_t *dev, spi_t spi, gpio_t cs);
+
+/**
+ * @brief Reset the TMP006 sensor. After that, the sensor should be reinitialized.
```
TMP006 -> ADT7310 I guess.

LudwigOrtmann commented on drivers/include/adt7310.h:
```diff
@@ -0,0 +1,164 @@
+ *
+ * @param[in]  raw          raw ADC value
+ *
+ * @return                  floating point representation of temperature in degrees celsius
```
celsius -> Celsius

gebart commented:
@LudwigOrtmann Addressed comments

PeterKietzmann commented on drivers/include/adt7310.h:
```diff
@@ -0,0 +1,154 @@
+ *
+ * @param[out] dev          device descriptor of sensor to initialize
+ * @param[in]  spi          SPI bus the sensor is connected to
+ * @param[in]  cs           GPIO pin which the chip select signal is connected to
```
is "which" correct here?

PeterKietzmann commented on drivers/adt7310/adt7310.c:
```diff
@@ -0,0 +1,190 @@
+    gpio_init_out(dev->cs, GPIO_NOPULL);
+    gpio_set(dev->cs);
+
+    for (int i = 0; i < 8; ++i) {
```
Is this needed or "just" for debug? Just wondering...

gebart replied:
yes, some left over from development, sorry. I've put it inside `#if ENABLE_DEBUG` now.

PeterKietzmann commented on drivers/include/adt7310.h:
```diff
@@ -0,0 +1,154 @@
+typedef struct {
+    spi_t spi;              /**< SPI bus that the sensor is connected to */
+    gpio_t cs;              /**< CS pin GPIO handle */
+    bool initialized;       /**< sensor status, true if sensor is initialized */
```
I think like that

PeterKietzmann commented:
Looks valid to me at first sight.

haukepetersen commented on drivers/include/adt7310.h:
```diff
@@ -0,0 +1,154 @@
+ *
+ * @return                  floating point representation of temperature in degrees Celsius
+ */
+float adt7310_convert(int16_t raw);
```
just a suggestion: how about changing the interface to 
```c
int16_t adt7310_read_raw(dev); // return raw value
int16_t adt7310_read(dev);     // return value in m°C
and
float adt7310_read_float(dev); // return value in °C
```
The error conditions you could either omit (just return 0 or something) or return INT16_MIN or similar. Opinions on this?

gebart replied:
good idea, however it will have to be `int32_t adt7310_read(dev);` in order to represent temperatures above 32.767 C

haukepetersen replied:
oh right :-)

gebart commented:
I updated the code but I think I might have broke my sensor or some wire, every SPI transfer comes back with 0xff.. I'll post back when I have tested this.

haukepetersen commented:
Very nice. Code looks good to me. The only thing I am not sure if I like it, is the external initialization of the SPI port (I know this in general is subject to ongoing discussion). IMHO this (i) breaks the consistency with all other existing drivers and (ii) does make it harder to use a device. Actually what I like to prevent is the situation, that you have to look into the initialization code/doc everytime  before you use a driver to find out which of the used peripherals you have to initialize beforehand... Also in the end, the driver knows best, which parameters to use (e.g. speed, etc.).

kaspar030 commented:
+1 for moving spi initialization into the driver.

haukepetersen commented:
@gebart: how about moving the initialization into the driver for know and we give the general discussion a push to decide how we handle this situation in general for the future?

gebart commented:
@haukepetersen I can live with that for the time being until a consensus is reached in #2528

haukepetersen commented:
I can if you document it properly for the `init` function :-)

ACK when this is done.

haukepetersen commented:
@gebart: I don't really understand my own last comment anymore... Where were we at with this PR? If you could either add documentation to the init function, that the SPI bus needs to be pre-initialized or if you could move the SPI init into the driver init, you can squash and we can merge this.

gebart commented:
@haukepetersen 
 - Added the comment about pre-initialized SPI bus to the documentation for adt7310_init.
 - Rebased after the new GPIO API was merged.

Tested on real hardware on my side. Waiting for Travis.

LudwigOrtmann commented:
pre-ACKed, and go =)
